### ISSUE TITLE
Using NPM SoftAP Setup module?

### ISSUE THREAD

**Issue by affinity-matrix** [msolters/softap-setup-meteor#1]
Hey!

This is super awesome! I noticed you're using the browserified 'hardcopy' of the SoftAP Setup module, and I wanted to see what I could do to help make it easier to use it straight from NPM. I know we're already talking about updating it to be more browser-friendly, but I'm not super familiar with Meteor apps, so I was curious to know if there's anything else we can do to make it easy to use in Meteor.

Thanks! :smile_cat:

**Comment by msolters** [comment]
Yeah so there's a couple threads being pulled together here.

For browser integration there are two main issues:
-  It seems not all browsers want to play nice with SoftAP.
-  In addition, some web assets get wonky or disappear when you make web-based setup interfaces that load content _after_ disconnecting from the server hosting the interface.  Like if I try to get the font awesome "lock" to indicate a secure network in my scanned AP results list, since I'm connected to the Photon, that resource doesn't resolve, and I get a square. No good.

**So I wanted to move towards a thin, downloadeable native client if I could.**

So let me answer you in two parts.

# Browser-based setup

**TLDR**: Meteor works fine with NPM modules so long as those modules are browserifiable.

Currently, what I've deployed to http://photonsoftap.meteor.com will work (in some browsers like Chrome), but it requires the user to manually toggle the WiFi connection between their router and the Photon.  That's fine for most users who aren't like 50+.  We certainly don't want to get rid of that.  So, what can you guys do to facilitate browser implementation?

It's all about the browserification.  Depending on your JS framework (or lack thereof) there are multiple ways to do this.  But to begin, the underlying code _must_ be browserifiable.  **So first of all, accept [brewnerd](https://github.com/brewnerd)'s PR**.  Publish that codebase to NPM.  Then you'll have an NPM module that users can browserify if and however they want to.

In vanilla JS, there exist NPM tools to do this, and then you end up just with some minified JS file that you include and there you go.

But in Meteor, there [exist packages](https://atmospherejs.com/cosmos/browserify) that let users simply declare NPM dependencies, and specify they want them to be automatically browserified.  As long as the NPM module is browserifiable, as brewnerd's PR is, that'll work automagically.

# Native

In this sense Meteor is a natural choice, because I can use the [electrify](https://atmospherejs.com/arboleya/electrify) package to instantly make a self-contained .exe, .app or linux executable package of my web app that gets delivered in its own tiny version of Chromium.  That's what the latest commit of my repo here is doing.  I've tested it in Windows, MacOS and Linux.  No compatibility issues.

Now, **if** we're using a native client, the server===client, and we have access to their hardware.  So, I've made an NPM package that allows us to programatically scan for WiFi, as well as connect/disconnect to access points:
-  [NPM Package](https://www.npmjs.com/package/wifi-control)
-  [Repo](https://github.com/msolters/wifi-control-node)

So, I've turned this NPM package into a [Meteor package](https://atmospherejs.com/msolters/wifi-control).

Now, if you want me to wrap a Meteor package for your NPM module, I can easily do that as well.  But it's generally recommended that you just go ahead and use something like [the npm package](https://atmospherejs.com/meteorhacks/npm) and pull the source in with no middle wrapper -- unless its something like accessing the WiFi card, in which case you need to add some Meteor-specific sugar to extend that server-only capability to the client somehow.  We can talk more about that.

---

Now, this app, as it stands, will pull in my `wifi-control` NPM module.  So when it boots up, it scans the nearby access points, and returns a list that contain "Photon-".  When you select one of those, it automatically instructs your operating system to connect to that network, and then it proceeds with using the `SAP` object to negotiate with the Photon.  So it's a local web app that has the same exact behaviour as say the Particle Android app.  (Note: iOS does not allow automatic WiFi configuration)

But, it's delivered as a 100Mb or so .exe, .app or executable Linux script.

I'm currently working on the wifi-control module (it's hard to test on Windows as I don't have it) but I think it will provide a much more reliable and automated setup than in the browser.

**Comment by msolters** [comment]
BTW that ~100Mb is including a local MongoDB and a bunch of other stuff that I never use in the setup app.  I'm sure we can make this smaller by just making a native Electron app, but then we lose the elegance of Meteor's templating and event handling.  I'm sure there are many in the community that would prefer to roll their own setup this way.  But the wifi-control and SoftAP NPM modules are what they will ultimately depend on to do the heavy lifting.

**Comment by affinity-matrix** [comment]
@msolters thank you! this is awesome input. I'm going to try to digest all of this, but everything you said sounds amazing. I had initially been trying to work with PhoneGap/Cordova and Ionic, but Angular is terrible, and I hadn't even thought of Meteor since 0.4.2 days (when it was kind of not that great). I'm super stoked to realize that it's so powerful now. Right now I'm working on a refactor of our JS client library which will incorporate the SoftAP Setup functionality, which will be easily browserified. 

I'd also love to chat some more about Meteor after I've got this refactor published, and perhaps you will have some input on that as well.

Thanks!

**Comment by msolters** [comment]
Hey, I saw you merged the PR, which is great!

I've been testing the ability to browserify the new module via Meteor's package system, but I've run into a roadblock: I can't test npm code until it's been published to npm. 

As [indicated here](https://github.com/elidoran/cosmos-browserify/issues/13), Meteor's browserification is relying on npm's build process and I can't actually test with a repo.

What kind of time frame are we looking at to push this new PR to npm?

**Comment by affinity-matrix** [comment]
@msolters I just published it like 5 minutes ago as version 3.0.0! I will tag it as a release here shortly: https://github.com/spark/softap-setup-js/releases

Let me know if I can do anything else to help w/ testing. I'm also planning on continuing to work with @lourd on an ES2015 version of the library which should be nice to work with within the Meteor ecosystem as well, so I'd love to have your input on that as far as the context of Meteor is concerned.

**Comment by msolters** [comment]
OK, that's great.  Thanks!  I've already tested against Meteor -- this version **doesn't** work, but the fix is simple.  I've sent you a PR.

If you could accept it, and publish maybe like a v3.1.0 or what have you to NPM, your library will be ready for consumption by the Meteor community with 0 problems :100: 

Let me know when you do -- I will post an example Meteor app in the forums.
